rest-client/src/main/java/org/sagebionetworks/bridge/rest/UserSessionInfoProvider.java

```diff
@@ -1,76 +1,99 @@
-        this.authenticationApi = retrofit.create(AuthenticationApi.class);
-        this.sessionInterceptor = sessionInterceptor;
+    UserSessionInfoProvider(AuthenticationApi authenticationApi, SignIn signIn) {
+        checkNotNull(authenticationApi);
+        checkNotNull(signIn);
+        
+        this.authenticationApi = authenticationApi;
+        this.signIn = signIn;
     }
-
+    
+    synchronized UserSessionInfo getSession() {
+        return session;
+    }
+    
     /**
-     * Retrieves a session. Tries to retrieve from a cache before calling a service.
+     * Retrieves a session. If a session hasn't been captured by an interceptor, this provider will attempt 
+     * to retrieve it by signing in.
      *
-     * @param signIn
-     *         sign in credentials for a user
-     * @return session info for the user, or null if the return value was not a 200
+     * @return session info for the user
      * @throws IOException problem while signing in
      */
-    public UserSessionInfo retrieveSession(SignIn signIn) throws IOException {
-        UserSessionInfo session = retrieveCachedSession(signIn);
-
+    synchronized UserSessionInfo retrieveSession() throws IOException {
         if (session != null) {
             return session;
         }
-
-        LOG.debug("No session, call intercepted for authentication attempt: " + signIn.getEmail());
-        authenticationApi.signIn(signIn).execute();
-
-        // Calling sign in will cause the session to be saved by the session interceptor. No further
-        // work is required here.
-        return sessionInterceptor.getSession(signIn);
+        // There's no session, so our only option is to sign in again.
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(debugString("null")+ " signing in");
+        }
+        signIn();
+        return session;
     }
-
-    /**
-     * Retrieves a session without making a network call.
-     *
-     * @param signIn
-     *         sign in credentials for a user
-     * @return session info for the user
-     */
-    public UserSessionInfo retrieveCachedSession(SignIn signIn) {
-        if (signIn == null) {
-            return null;
+    
+    synchronized void setSession(UserSessionInfo session) {
+        if (LOG.isDebugEnabled()) {
+            String token = (session==null) ? "null" : session.getSessionToken();
+            LOG.debug(debugString(token));
         }
-
-        return sessionInterceptor.getSession(signIn);
+        this.session = session;
     }
-
-    /**
-     * Removes a session from the cache. This should be called if a session returned by this instance turns out to be
-     * no longer valid.
-     *
-     * @param session the user's session
-     */
-    void purgeCachedSession(UserSessionInfo session) {
-        sessionInterceptor.removeSession(session);
+    
+    synchronized void reauthenticate() throws IOException {
```

Comment:
I like this refactor, really cleans things up. And centralizes where credential handling happens (authHandler and sessionInterceptor no longer need to know about signIn, reauth token, etc)